src/transformers/cache_utils.py

```diff
@@ -320,3 +320,361 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        self.num_blocks = num_blocks
+        self.block_size = block_size
+        self.key_cache: Dict[int, Tuple[torch.Tensor]] = {}  # layer_idx -> key_cache
+        self.value_cache: Dict[
+            int, Tuple[torch.Tensor]
+        ] = {}  # layer_idx -> value_cache
+        self.seen_tokens = (
+            0  # Used in `generate` to keep tally of how many tokens the cache has seen
+        )
+
+        # cache runtime management information
+        self.free_blocks = list(range(num_blocks))  # free blocks
+        self.block_ref_count = [
+            0
+        ] * self.num_blocks  # init the reference count for each physical block
+        self.block_tables = (
+            {}
+        )  # mapping logical block to physical blocks for each sequence
+        self.context_lens = {}  # context length for each sequence
+
+        # The follow two states are shared accross layer but only for the current decode step. Need to update for every decode step.
+        self.batch2seq = (
+            {}
+        )  # mapping batch index to {seq_id0, seq_id1, ...} to enable prompt sharing.
+        self.slots_mapping = []  # mapping logical slots to physical slots.
+
+    @classmethod
+    def from_legacy_cache(
+        cls, past_key_values: Optional[List[torch.FloatTensor]]
+    ) -> "PagedAttentionCache":
+        if past_key_values is None:
+            return cls()
+        cache = cls()
+        for layer_idx, (key_states, value_states) in enumerate(zip(*past_key_values)):
+            cache.update(key_states, value_states, layer_idx)
+        return cache
+
+    def copy_on_write(self, src_block_idx: int, dst_block_idx: int):
+        """
+        Copy the content of src_block_idx to dst_block_idx.
+
+        Args:
+            src_block_idx (int): The index of the source block.
+            dst_block_idx (int): The index of the destination block.
+        """
+        for layer_idx in range(len(self.key_cache)):
+            self.key_cache[layer_idx][dst_block_idx] = self.key_cache[layer_idx][
+                src_block_idx
+            ].clone()
+            self.value_cache[layer_idx][dst_block_idx] = self.value_cache[layer_idx][
+                src_block_idx
+            ].clone()
+
+    def allocate(self, seq_idx: int, key_len: int, context_len: int) -> List[int]:
+        """
+        Allocate physical slots for a given sequence index, key length and context length.
+
+        Args:
+        - seq_idx (int): The index of the sequence.
+        - key_len (int): The length of the key.
+        - context_len (int): The length of the context.
+
+        Returns:
+        - slots (list): The physical slots for the given sequence, where 1 slot is allocated for 1 token state.
+        """
+        slots = []
+        if seq_idx not in self.block_tables:
+            # allocate blocks for this sequence
+            assert context_len == 0
+            needed_blocks = (key_len + self.block_size - 1) // self.block_size
```

Comment:
Shouldn't `needed_blocks` be the smallest number of blocks where the current cache would fit, i.e. `key_len/self.block_size` rounded up? (i.e. `-(-key_len // self.block_size)`)

Same comment for `target_blocks` in the `else` branch.

If my comment is wrong, what am I missing?

Reply:
Thanks  @gante I will refine it according to your comments.

Reply:
Yes. you are right. the needed_blocks and targed_blocks are the smallest number of blocks needed for the entire context(past key/value + current key/value). For the _if_ branch, it is a new sequence, and the past context is 0. For the else branch, the past context len is more than 0, and there may be free slots in the last block of this sequence, we need to firstly use these free slots before allocating new physical blocks. 
The _"(key_len + self.block_size - 1) // self.block_size"_  is same to _-(-key_len // self.block_size)_ to do rounded up.